domain/model/service/state_test.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/credential"
+	"github.com/juju/juju/core/life"
+	coremodel "github.com/juju/juju/core/model"
+	"github.com/juju/juju/core/user"
+	coreuser "github.com/juju/juju/core/user"
+	usererrors "github.com/juju/juju/domain/access/errors"
+	clouderrors "github.com/juju/juju/domain/cloud/errors"
+	"github.com/juju/juju/domain/model"
+	modelerrors "github.com/juju/juju/domain/model/errors"
+)
+
+type dummyStateCloud struct {
```

Comment:
Seeing all of these hand crafted dummy structs reminds me why we decided to shift away from this approach a few years back and start using gomock. I still wish we used a consistent approach. There's now a mix of this and gomock in the various service packages.

Reply:
We can discuss this till the cows come home. Gomock has its place I agree with that. But at some point you have to be honest with yourself and say I am using Gomock because I don't want to write some more code and test things properly.

We are forming a very important business layer contract here and when we use Gomock we aren't testing anything. We assert internal implementation and not the pre and post conditions. How can a developer having never understood the business logic that goes into model have confidence in their changes when Gomock is involved. If we solve that problem I am happy to use them in this case.

But fundamentally this argument just keeps coming back to I hate writing more code and so we take the cheat way out. I care about the business logic formed at the service layer and that it is correct. You can't assert that long term with Gomock.

Reply:
It's not about not wanting to write more code. We used the hand crafted approach for a long time and as a team have moved away from it when gomock was introduced. I just am upset at the lack of consistency across developers. We need to agree on a standard for the team and stick to it.

Reply:
I agree as per above; standardising on mocks is the way to go here.